beacon-chain/rpc/beacon_server.go

```diff
@@ -210,18 +210,81 @@ func (bs *BeaconServer) PendingDeposits(ctx context.Context, _ *ptypes.Empty) (*
+	allDeps := bs.beaconDB.AllDeposits(ctx, bNum)
+	if len(allDeps) == 0 {
+		return &pb.PendingDepositsResponse{PendingDeposits: nil}, nil
+	}
+
+	// Need to fetch if the deposits up to the state's latest eth 1 data matches
+	// the number of all deposits in this RPC call. If not, then we return nil.
+	beaconState, err := bs.beaconDB.State(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch beacon state: %v", err)
+	}
+	h := bytesutil.ToBytes32(beaconState.LatestEth1Data.BlockHash32)
+	_, latestEth1DataHeight, err := bs.powChainService.BlockExists(ctx, h)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch eth1data height: %v", err)
+	}
+	upToLatestEth1DataDeposits := bs.beaconDB.AllDeposits(ctx, latestEth1DataHeight)
+	if len(upToLatestEth1DataDeposits) != len(allDeps) {
+		return &pb.PendingDepositsResponse{PendingDeposits: nil}, nil
+	}
+	depositData := [][]byte{}
+	indices := []uint64{}
+	for i := range upToLatestEth1DataDeposits {
+		depositData = append(depositData, upToLatestEth1DataDeposits[i].DepositData)
+		indices = append(indices, upToLatestEth1DataDeposits[i].MerkleTreeIndex)
+	}
+
+	depositTrie, err := trieutil.GenerateTrieFromItems(depositData, int(params.BeaconConfig().DepositContractTreeDepth))
+	if err != nil {
+		return nil, fmt.Errorf("could not generate historical deposit trie from deposits: %v", err)
+	}
+	for i := range pendingDeps {
+		proof, err := depositTrie.MerkleProof(int(pendingDeps[i].MerkleTreeIndex))
+		if err != nil {
+			return nil, fmt.Errorf(
+				"could not generate merkle proof for deposit at index %d: %v",
+				pendingDeps[i].MerkleTreeIndex,
+				err,
+			)
+		}
+		// For every deposit, we construct a Merkle proof using the powchain service's
+		// in-memory deposits trie, which is updated only once the state's LatestETH1Data
+		// property changes during a state transition after a voting period.
+		pendingDeps[i].MerkleBranchHash32S = proof
+	}
+	return &pb.PendingDepositsResponse{PendingDeposits: pendingDeps}, nil
 }
 
 func (bs *BeaconServer) defaultDataResponse(ctx context.Context, currentHeight *big.Int, eth1FollowDistance int64) (*pb.Eth1DataResponse, error) {
-	ancestorHeight := currentHeight.Sub(currentHeight, big.NewInt(eth1FollowDistance))
+	ancestorHeight := big.NewInt(0).Sub(currentHeight, big.NewInt(eth1FollowDistance))
 	blockHash, err := bs.powChainService.BlockHashByHeight(ctx, ancestorHeight)
 	if err != nil {
 		return nil, fmt.Errorf("could not fetch ETH1_FOLLOW_DISTANCE ancestor: %v", err)
 	}
-	// TODO(#1656): Fetch the deposit root of the post-state deposit contract of the block
-	// references by the block hash of the ancestor instead.
-	depositRoot := bs.powChainService.DepositRoot()
+	// Fetch all historical deposits up to an ancestor height.
+	allDeposits := bs.beaconDB.AllDeposits(ctx, ancestorHeight)
+	depositData := [][]byte{}
+	// If there are no historical deposits up to an ancestor height, then we just fetch the default
```

Comment:
`if there are no historical deposits up to an ancestor height` where is this logic implemented? 
I don't see any `if len(deposits) == 0` logic